File: Haskell/P8.hs
-------------------------------------------------------------------------------
-- Solved by OmarG247
-- Problem 8: Delete consecutive duplicates of elements in a list
-------------------------------------------------------------------------------

clean :: (Eq a) => [a] -> [a]
clean (a:b)
    = a : (clean' b a)
clean [] 
    = []

clean' :: (Eq a) => [a] -> a -> [a]
clean' (a:b) current
    | a /= current
        = a : (clean' b a)
    | otherwise 
        = clean' b current
clean' [] current
    = []

File: Prolog/P8.pl
%-----------------------------------------------------------------------------
% Solved by Walsker
% Problem 8: Delete consecutive duplicates of elements in a list
%-----------------------------------------------------------------------------

% Removes the given value from the front of list until there is one left
remove(_, [], []).
remove(A, [B|C], X) :- A == B, remove(A, C, X); X = [B|C].

% Solution
compress([], []).
compress([A|B], X) :- remove(A, B, Y), compress(Y, Z), X = [A|Z].